Utilize pre-commit, black, isort and doc8

tim-schilling opened the pull request:
It will be easier to contribute if the style preferences of the project were
enforced automatically. pre-commit is a tool that supports this. I've set it up
to apply black, isort and doc8 to standardize the coding style.

Obviously, this is a large change and I don't have a ton of involvement with the project historically. However, these tools have made my life easier on other projects, especially those that I am reviewing pull requests for.

Edit:
If this is too much to take on, that's understandable and I won't be offended.

medmunds commented:
Thanks for this. Black and isort have actually been on my "would be nice" list for a while, but never bubbled up to the top.

Probably the tox testenv:lint should also be updated to match. Or ideally just run the pre-commit hooks from tox -e lint (or vice versa) so that tox and pre-commit stay in sync.

I'd also like to use [.git-blame-ignore-revs](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) to make it easier to spot code changes around the mass reformatting. (So probably we'd want to split the reformatting into one rev that can be git-blame-ignored, and then add the pre-commit hooks in the next rev, which wouldn't be ignored. I think.)

tim-schilling commented:
@medmunds what do you think about dropping support for Django 2-3.1 and Python 3.5? I realized that I configured black to target 3.8 rather than the lowest possible version.

tim-schilling commented:
@medmunds I pulled out the style changes into a large commit and added that to `.git-blame-ignore-revs`. I've also removed the lint usage in tox in favor of setting up an integration with https://pre-commit.ci/ which is free for open source libraries. The CI portion of the integration will validate open PRs and enforce style guidelines.

medmunds commented:
Thanks again for this, and sorry for the delayed responses—I'm kind of swamped with some other matters lately.

I'll drop a handful of specific questions/comments into a code review.

> what do you think about dropping support for Django 2-3.1 and Python 3.5?

We can and should drop support for Django 2.x and Python 3.5. That pending deprecation was announced with [Anymail v8.6](https://github.com/anymail/django-anymail/blob/main/CHANGELOG.rst#v86-lts) on 2022-05-15.

I'd like to keep Django 3.0 and 3.1 support (on Python 3.6 or later) until it's time to drop support for all Django 3.x. (As a general rule, I try to retain support for Django versions a while after Django ends support, so that Anymail is never the package forcing a project to upgrade.)

[Dropping version support should be a separate PR. PR welcome, or I'll try to get to that this weekend.]

> removed the lint usage in tox

There are two goals for tox usage in this project: 1) If a developer runs `tox` locally, and the tests pass, they should be confident that their changes will also pass CI (with the possible exception of live integration tests). 2) Reduce or eliminate duplication of CI configuration. (We used to have a separate .travis-ci and tox.ini which were almost never in sync.)

So I'd like to keep the tox lint environment, and have it run the pre-commit hooks. I haven't worked with pre-commit before, but it seems like that would just be `pre-commit run --all-files`? (The pre-commit docs also have notes on [usage with tox](https://pre-commit.com/#usage-with-tox).)

As far as pre-commit.ci, I'd prefer to keep the CI all in GitHub Actions for now, just to minimize the number of service integrations that have to be maintained.

medmunds commented on .flake8:
```diff
@@ -1,2 +0,0 @@
-[flake8]
-max-line-length = 120
```
Also, in setup.py's `requirements_dev` need to drop flake8, add pre-commit. 

[Long-term, I'd really like to modernize this project's packaging, which would eliminate a lot of setup.py. But that's a whole separate topic.]

tim-schilling replied:
I updated requirements_dev in setup.py

medmunds commented:
(I've dropped Python 3.5 and Django 2.x support on the main branch, so black can target Python 3.6.)

tim-schilling commented:
@medmunds thank you for the quality review. I haven't forgotten about this. My OSS availability is weaning a bit, but I'll try to address this over the weekend.

medmunds commented:
@tim-schilling I completely understand—no rush.

tim-schilling commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,34 @@
+    hooks:
+    -   id: black
+        language_version: python3
+        entry: black --target-version=py36
```
It looks like black officially only supports configuration via a pyproject.toml file. It also supports a .black file, but that's not what you want. The project would need to switch to something like hatch, flit or poetry for packaging to drop the setup.cfg file.

tim-schilling commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,34 @@
+-   repo: https://github.com/pycqa/isort
+    rev: 5.10.1
+    hooks:
+    -   id: isort
```
Yes, I have the isort profile set in setup.cfg

tim-schilling replied:
I did have isort somewhat configured in setup.cfg. I'll move over the entire configuration.

tim-schilling replied:
This should be complete. I had to rework the commits due to the rst rule change.

tim-schilling commented on .flake8:
```diff
@@ -1,2 +0,0 @@
-[flake8]
-max-line-length = 120
```
This was moved to setup.cfg so nothing should need to be updated yet.

Django's .editorconfig didn't have anything for *.rst, but I'll add a line length config for ours.

tim-schilling commented on docs/contributing.rst:
```diff
@@ -51,8 +51,7 @@ Pull requests
 * Please include test cases.
-* We try to follow the `Django coding style`_
-  (basically, :pep:`8` with longer lines OK).
+* Install `pre_commit`_ for linting.
```
This should be handled, though I'll need you to re-review the new messaging.

tim-schilling commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,34 @@
+    -   id: python-check-blanket-noqa
+    -   id: python-no-eval
+    -   id: python-no-log-warn
+    -   id: rst-backticks
```
I removed that rst-backticks check. It drastically reduced the number of doc file changes as you expected.

medmunds commented:
@tim-schilling sorry for the lengthy delay getting back to this. I've "rebased" to the latest main, and made a few adjustments and additions in the formatting config.

Decided to locate all the tool config in pyproject.toml rather than setup.cfg, in preparation for modernizing Anymail's packaging at some point in the (hopefully near) future. Exception is flake8, which doesn't read pyproject.toml—so just left that in .flake8.

Thanks again for doing the legwork on this.

tim-schilling commented:
Thank you for maintaining the package and seeing this one through!
